tests/examples/shims/root.lua

```diff
@@ -1,9 +1,126 @@
 local root = {_tags={}}
 
+local hotkeys = nil
+
 function root:tags()
     return root._tags
 end
 
+function root:size() --TODO use the screens
+    return 0, 0
+end
+
+function root:size_mm()
+    return 0, 0
+end
+
+function root.cursor() end
+
+-- GLOBAL KEYBINDINGS --
+
+local keys = {}
+
+-- Note: the tests are not expected to create 1000 of them, so to keep the
+-- implementation minimal, it doesn't try to hash them and simply loop over
+-- every entry.
+function root.keys(k)
+    keys = k or keys
+    return keys
+end
+
+-- FAKE INPUTS --
+
+local conversion, mods = {
+    Super_L   = "Mod4",
+    Control_L = "Control",
+    Shift_L   = "Shift",
+    Alt_L     = "Mod1",
+    Super_R   = "Mod4",
+    Control_R = "Control",
+    Shift_R   = "Shift",
+    Alt_R     = "Mod1",
+}, {}
+
+local function add_modkey(key)
+    if not conversion[key] then return end
+    mods[conversion[key]] = true
+end
+
+local function get_mods()
+    local ret = {}
+
+    for mod in pairs(mods) do
+        table.insert(ret, mod)
+    end
+
+    return ret
+end
+
+local function remove_modkey(key)
+    if not conversion[key] then return end
+    mods[conversion[key]] = nil
+end
+
+local function match_modifiers(mods1, mods2)
+    if #mods1 ~= #mods2 then return false end
+
+    for _, mod1 in ipairs(mods1) do
+        local found = false
+        for _, mod2 in ipairs(mods2) do
+            if mod1 == mod2 then
+                found = true
+                break
+            end
+        end
+        if not found then return false end
+    end
+
+    return true
+end
+
+local function execute_press_keybindings(key, event)
+    -- It *could* be extracted from gears.object private API, but it's equally
+    -- ugly as using the list used by the hotkey widget.
+    if not hotkeys then
+        hotkeys = require("awful.key").hotkeys
+    end
+
+    for _, v in ipairs(hotkeys) do
+        if key == v.key and match_modifiers(v.mod, get_mods()) and v[event] then
+            v[event]()
+            return
+        end
+    end
+end
+
+local fake_input_handlers = {
+    key_press      = function(key)
+        add_modkey(key)
+        if keygrabber._current_grabber then
+            keygrabber._current_grabber(get_mods(), key, "press")
+        else
+            execute_press_keybindings(key, "press")
+        end
+    end,
+    key_release    = function(key)
+        remove_modkey(key)
+        if keygrabber._current_grabber then
+            keygrabber._current_grabber(get_mods(), key, "release")
+        else
+            execute_press_keybindings(key, "release")
```

Comment:
What exactly does the `press` in the name of the function refer to? How about just `execute_keybindings`?